add extremesdim

PythagoreanCult opened the pull request:
Closes #5

PythagoreanCult committed:
Add files via upload

PythagoreanCult committed:
Delete FractalDimensions.jl

PythagoreanCult committed:
add extremesdim

PythagoreanCult committed:
changed misspelled persistence

PythagoreanCult committed:
remove annoying folder
 Please enter the commit message for your changes. Lines starting

PythagoreanCult committed:
Merge branch 'Ignacio'

Datseris reviewed (commented):
Thanks for the contribution. This code will require substantial refactoring. Some lines of code has subpar peformance, which is no surprise given that it was adapted from some matlab code with terrible performance. I will go through the PR and leave comments. However, it is beneficial for you to also read through the Julia performance tips here: https://docs.julialang.org/en/v1/manual/performance-tips/ 

Specifically, you are creating a lot of new arrays without need. For example, every time you write `q.*Si` this creates a new array and allocates memory for it. But you don't actually need to store this array as is evident by the later code.

The other thing you should have a look at is the Neighborhood.jl package: https://julianeighbors.github.io/Neighborhood.jl/dev/ which is used in the rest of the FractalDimensions.jl file to optimize nearest neighbor searches. For example you can look at the vanilla correlation sum implementation for how this package is used. 

Lastly, **the most important thing is that you need to add tests for this PR**. Go into the `test` folder, create a new file, add tests for the new functionality, and include the new test file in runtests.jl.

---

And don't worry if this is too much and you can't do it, I'll do them in due time, and I'll also leave a more detailed review in due time.

Datseris commented on src/extremes_based/extremesdim.jl:
```diff
@@ -0,0 +1,126 @@
+"""
+function loc_dimension_persistence(x:: AbstractVector, quanti:: Real)
+
+    println("Computing dynamical quantities")
```
If you want to print a progress bar here, then please use the ProgressMeter.jl code as we do in e.g., the vanilla correlation sum source code.

Datseris commented on src/extremes_based/extremesdim.jl:
```diff
@@ -0,0 +1,126 @@
+    inverse of the persistence of the extremes around that point.
+    
+"""
+function loc_dimension_persistence(x:: AbstractVector, quanti:: Real)
```
replace `quanti` with `q`

Datseris replied:
I am confused here. `x` is an Abstract vector? But you say it is a trajectory... Did you mean to write `x::AbstractStateSpaceSet`?

Datseris commented on src/extremes_based/extremesdim.jl:
```diff
@@ -0,0 +1,126 @@
+"""
+loc_dimension_persistence(x:: AbstractVector, quanti:: Real) -> D1, θ
+    Computation of the local dimensions D1 and the extremal index θ for each observation in the 
+    trajectory x, for a given quantile "quanti". The extremal index can be interpreted as the 
```
When you refer to things in code, typeset them in code. Don't write "quanti", write `quanti`.

Datseris commented on src/extremes_based/extremesdim.jl:
```diff
@@ -0,0 +1,126 @@
+
+"""
+loc_dimension_persistence(x:: AbstractVector, quanti:: Real) -> D1, θ
+    Computation of the local dimensions D1 and the extremal index θ for each observation in the 
```
The spacing is wrong here. The function call signature should be idented 4 spaces, while the docstirng should have no identation.

Datseris commented on src/extremes_based/extremesdim.jl:
```diff
@@ -0,0 +1,126 @@
+        #The local dimension is the reciprocal of the exceedances of the PoTs
+        D1[j] = 1 ./ mean(logextr .- thresh);
+    end
+    return [D1, θ]
```
Why are you returning things as a vector? Just `return D1, θ`.

Datseris commented on src/extremes_based/extremesdim.jl:
```diff
@@ -0,0 +1,126 @@
+    θ = zeros(size(x[:,1]));
+    for j in range(1,length(D1))
+        # Compute the observables
+        logdista = -log.([euclidean(x[j,:],x[i,:]) for i in range(1,length(x[:,1]))]);
```
If `x` is a state space set, jsut write `x[j]` and `x[i]` to get the i and j data points.

Datseris commented on src/extremes_based/extremesdim.jl:
```diff
@@ -0,0 +1,126 @@
+    Li = findall(x -> x > u, Y);
+    Ti = diff(Li); 
+    Si = Ti .- 1;
+    Nc = length(findall(x->x>0,Si));
```
`findall` allocates a vector in each call. Instead of `length(findall(x->x>0,Si));` write `count(x -> x>0, Si)`.

Datseris commented on src/extremes_based/extremesdim.jl:
```diff
@@ -0,0 +1,126 @@
+        error("The number quanti represents a quantile and has to be between 0 and 1")
+    end
+
+    q = 1 - quanti;
```
Why are you ending each line here with `;`? Please don't do that in Julia, it isn't necessary.

Datseris commented on src/extremes_based/extremesdim.jl:
```diff
@@ -0,0 +1,126 @@
+        error("The number quanti represents a quantile and has to be between 0 and 1")
+    end
+
+    D1 = zeros(size(x[:,1]));
```
Please don't put `;` at the end of each command unless there is a reason to, i.e., there is a second command in the same line.

Datseris commented on src/extremes_based/extremesdim.jl:
```diff
@@ -0,0 +1,126 @@
+    θ = zeros(size(x[:,1]));
+    for j in range(1,length(D1))
+        # Compute the observables
+        logdista = -log.([euclidean(x[j,:],x[i,:]) for i in range(1,length(x[:,1]))]);
```
I confused about this line. What are you trying to do with `range(1,length(x[:,1]))`? This is such a weird syntax. What is this supposed to be?

Datseris replied:
Is `x` a StateSpaceSet? then you probably just want `length(x)`.

Datseris commented on src/extremes_based/extremesdim.jl:
```diff
@@ -0,0 +1,126 @@
+    a time series Y, given the quantile p and the corresponding threshold p.
+    
+"""
+function extremal_index_sueveges(Y:: AbstractVector, u:: Real, quanti:: Real)
```
This function can become much more compitationally efficient if it is replaced by a `for` loop instead of making everything vectorized via `diff` etc. You can do `continue` if current element of `x` isn't greater than `u`.

Datseris commented on src/extremes_based/extremesdim.jl:
```diff
@@ -0,0 +1,126 @@
+    Extremes, 10.1-2, 41-55, doi: 10.1007/s10687-007-0034-2
+     
+
+"""
```
Don't put this as a string, put it as a comment. Additionally, don't say what to cite here. Instead, put the citation properly in the docstring of the function below, as we do with the rest of the functions in the library with the `[^NameYear]` syntax.

Datseris reviewed (changes requested):
@pythagoreancult thanks for your contribution. 

In this review, I only leave minor comments. I am certain that the algorithm itself can be re-written for huge performance gains, but we need to resolve the minor comments first to understand what is going on...

What is crucial for me is for you to provide a test, or an example file, something that shows how this function is used, and with what kind of data. The documentation string says that `x` is a trajectory, but this is an English word, not a Julia word. What is it in Julia? Is it a vector of vectors, where each point is an observation? Is it a `StateSpaceSet`?

Depending on how much availability you have to work on this, I will also contribute directly to the PR, because I intend to use this code in an ongoing comparison of various fractal estimation methods.

Datseris committed:
turn the initial string into comment

Datseris committed:
change the return type to Tuple

This is standard practice; functions input and output tuples. There is no reason to make the returned Tuple into a Vector, and if anything it is bad practice; it makes a type unstable vector and it serves no purpose. Anything you could do with this two element vector you could do better and faster with a Tuple.

Datseris committed:
Delete all `;`

No instance of `;` in the source code had reason to exist. Julia is capable of automatically deducing when a statement ends. Only reason for `;` is to have to commands on one line like:
`x = 3; y = x + 2`.

Datseris committed:
organize import and export statements

Datseris committed:
Correct structure in docstrings

Datseris committed:
replace quanti with q

Also, make the error messages much shorter

Datseris committed:
remove space from all `:: ` 

Apparently it is possible to have space between `::` but it is very un-idiomatic; noone does it in real code, which makes putting spaces hard to read.

Datseris committed:
replace `AbstractVector` with `AbstractStateSpaceSet`

This is arguably the most important change I do in this PR. Read the docstring of `StateSpaceSet` to understand why.

Datseris committed:
Fix the `extremal_index_sueveges` function...

There were a lot of problems:

- the function name was different from where it was called
- the function input arguments were named differently from where the function was called. This was confusing.
- The order of arguments and the documentation string where wrong. This was ULTRA confusing. The threshold and quantile arguments were exchanging places in the documentation.
- The quantile and threshold were named the same way. Not possible to deduce which was what besides going through the algorithm.

Datseris committed:
rename variables for clarity

Datseris committed:
don't call extremal index; function is wrong

Datseris committed:
use a normal way to get length of set

Datseris committed:
add convenience function for dim

Datseris committed:
first pass at performance: complete refactor of functions

The point is: local estimation should be its own function because it is a dedicated piece of algorithm. Then, we can optmize stuff by looping over each "column" of the distance matrix. If it fits in memory, we also gain massive BLAS-enabled performance. If not, we can still optimize by pre-allocating vectors and not re-allocated memory in each pass.

Datseris committed:
add convenience function for local values

Datseris committed:
add progress bar

and also stop using `AbstractStateSpaceSet` in function call signatures

Datseris committed:
add test file

Datseris committed:
correctly use greater-equals instead of only greater

Datseris committed:
add second estimator for pareto sigma

Datseris committed:
parallelize via threads

Datseris committed:
Merge branch 'main' into Ignacio

Datseris committed:
Merge branch 'main' into Ignacio

Datseris committed:
add preliminary tests for extreme value dim

Datseris committed:
fix test for extremal index

Datseris committed:
wip] add type for deciding estimator (?)

Datseris committed:
we don't need a type in the end.

Datseris committed:
Merge branch 'main' into Ignacio

Datseris committed:
rename q to p as it quantifies a probability, not the quantile

Datseris committed:
wip evt

Datseris committed:
finish docs EVT

Datseris committed:
add passing tests to EVT

Datseris committed:
add to docs

Datseris committed:
changelog
